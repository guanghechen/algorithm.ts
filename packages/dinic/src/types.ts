import type { IDigraphEdge } from '@algorithm.ts/types'

export interface IDinicEdge extends IDigraphEdge {
  /**
   * The starting node of the arc.
   */
  from: number
  /**
   * The ending node of the arc.
   */
  to: number
  /**
   * Capacity of the arc.
   */
  cap: number
  /**
   * Flow on the arc.
   */
  flow: number
}

export interface IDinic {
  /**
   * Initialize the Dinic algorithm.
   * @param source  the source node
   * @param sink    the sink node
   * @param n       the number of nodes
   */
  init(source: number, sink: number, n: number): void
  /**
   * Add an edge into the residual network.
   * @param from    the starting node
   * @param to      the ending node
   * @param cap     capacity
   */
  addEdge(from: number, to: number, cap: number): void
  /**
   * Calculate the maximum flow of the residual network.
   */
  maxflow(): number
  /**
   * Calculate the minium cut of the residual network.
   */
  mincut(): Array<Readonly<IDinicEdge>>
}
