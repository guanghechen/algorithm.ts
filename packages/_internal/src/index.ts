export * from './constant'
export * from './types'
