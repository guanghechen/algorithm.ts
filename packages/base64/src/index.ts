import { Base64 } from './base64'

export * from './base64'

const base64 = new Base64()

export const encode = base64.encode.bind(base64)
export const decode = base64.decode.bind(base64)
export const validate = base64.validate.bind(base64)
export default { encode, decode, validate }
