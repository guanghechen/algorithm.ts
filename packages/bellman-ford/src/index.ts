import { createCircularQueue } from '@algorithm.ts/circular-queue'

export interface IEdge {
  /**
   * The other end of the edge.
   */
  to: number
  /**
   * The cost of walking along this side.
   */
  cost: number
}

export interface IGraph {
  /**
   * The number of nodes in the graph. (0-index)
   */
  N: number
  /**
   * The source node. (0-index)
   */
  source: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
  /**
   * An array recording the shortest distance to the source point.
   */
  dist: number[]
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
  /**
   * Used to check if an element is already in the queue.
   */
  inq?: boolean[]
  /**
   * Record the number of times an element is enqueued,
   * used to check whether there is a negative cycle.
   */
  inqTimes?: number[]
}

const ZERO = 0
const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)
const Q = createCircularQueue<number>()

/**
 * The bellman-ford algorithm, optimized with queue.
 *
 * @param graph
 * @param options
 */
export function bellmanFord(graph: IGraph, options: IOptions = {}): boolean {
  const { N, source, edges, G, dist } = graph
  const { INF = DEFAULT_INF, inq = [], inqTimes = [] } = options

  if (dist.length < N) dist.length = N
  if (inq.length < N) inq.length = N
  if (inqTimes.length < N) inqTimes.length = N

  dist.fill(INF, 0, N)
  inq.fill(false, 0, N)
  inqTimes.fill(0, 0, N)

  Q.init(N + 1)
  Q.enqueue(source)
  dist[source] = ZERO

  while (Q.size() > 0) {
    const o: number = Q.dequeue()!
    inq[o] = false

    for (const idx of G[o]) {
      const edge: IEdge = edges[idx]
      if (dist[edge.to] > dist[o] + edge.cost) {
        dist[edge.to] = dist[o] + edge.cost
        if (!inq[edge.to]) {
          Q.enqueue(edge.to)
          inq[edge.to] = true

          // eslint-disable-next-line no-plusplus
          if (++inqTimes[edge.to] > N) return false
        }
      }
    }
  }
  return true
}

export default bellmanFord
