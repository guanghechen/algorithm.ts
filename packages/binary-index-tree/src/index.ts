export * from './tree1'
export * from './tree2'
export * from './types'
export * from './util'
