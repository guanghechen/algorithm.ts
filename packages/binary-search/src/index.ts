export * from './binary-search'
export * from './lower-bound'
export * from './upper-bound'
export * from './types'
