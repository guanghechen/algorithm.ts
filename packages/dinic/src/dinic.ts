import type { CircularQueue } from '@algorithm.ts/circular-queue'
import { createCircularQueue } from '@algorithm.ts/circular-queue'
import type { Dinic, DinicContext, DinicEdge } from './types'

export function createDinic(): Dinic {
  let source: number // The source point in a network flow
  let target: number // The sink in a network flow
  let n: number // The number of nodes in a network flow
  let m: number // The number of edges in a network flow (not including the reverse edges).
  let answer: number
  let edgeTot: number
  const cur: number[] = [] // The next edge number to be considered of the edges starting from the i-th node.
  const dist: number[] = [] // The distance from the source node to the i-th node.
  const edges: DinicEdge[] = []
  const G: number[][] = []
  const Q: CircularQueue<number> = createCircularQueue()
  return { init, addEdge, maxflow, solve }

  function init(
    _source: number,
    _target: number,
    _n: number,
    _m: number,
  ): void {
    source = _source
    target = _target
    n = _n
    m = m << 1
    answer = 0

    // Resize arrays.
    if (cur.length < n) cur.length = n
    if (dist.length < n) dist.length = n
    if (edges.length < m) edges.length = m
    if (G.length < n) G.length = n

    edgeTot = 0
    Q.init(n + 1)
    for (let i = 0; i < n; ++i) G[i] = []
  }

  function addEdge(from: number, to: number, cap: number): void {
    G[from].push(edgeTot)
    // eslint-disable-next-line no-plusplus
    edges[edgeTot++] = { from, to, cap, flow: 0 }

    G[to].push(edgeTot)
    // eslint-disable-next-line no-plusplus
    edges[edgeTot++] = { from: to, to: from, cap: 0, flow: 0 }
  }

  function maxflow(): number {
    while (bfs()) {
      cur.fill(0, 0, n)
      answer += dfs(source, Number.MAX_SAFE_INTEGER)
    }
    return answer
  }

  function solve(fn: (context: DinicContext) => void): void {
    const context: DinicContext = { edgeTot, dist, edges, G }
    fn(context)
  }

  function bfs(): boolean {
    // Initialize the dist array.
    dist.fill(-1, 0, n)

    Q.enqueue(source)
    dist[source] = 0
    while (Q.size() > 0) {
      const o = Q.dequeue()!
      for (const i of G[o]) {
        const e = edges[i]
        if (dist[e.to] === -1 && e.cap > e.flow) {
          dist[e.to] = dist[o] + 1
          Q.enqueue(e.to)
        }
      }
    }

    return dist[target] !== -1
  }

  function dfs(o: number, minFlow: number): number {
    if (o === target || minFlow === 0) return minFlow

    let flow = 0
    for (let g = G[o]; cur[o] < g.length; ++cur[o]) {
      const x = g[cur[o]]
      const e = edges[x]
      if (dist[e.to] === dist[o] + 1) {
        const f = dfs(e.to, Math.min(minFlow, e.cap - e.flow))
        if (f <= 0) continue
        e.flow += f
        edges[x ^ 1].flow -= f
        flow += f
        // eslint-disable-next-line no-param-reassign
        minFlow -= f
        if (minFlow === 0) break
      }
    }
    return flow
  }
}
