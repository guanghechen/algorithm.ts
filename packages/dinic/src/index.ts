export * from './dinic'
export * from './types'
