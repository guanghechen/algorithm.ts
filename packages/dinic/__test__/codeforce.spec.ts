import fs from 'fs-extra'
import { createReader, loadFixtures, locateFixtures } from 'jest.setup'
import path from 'path'
import { createDinic } from '../src'

describe('codeforces', function () {
  describe('contest 1082', function () {
    test('g', function () {
      const dinic = createDinic()
      function solve(content: string): number {
        const io = createReader(content)
        const [n, m] = io.readIntegersOfLine()

        const source = 0
        const target: number = n + m + 1
        dinic.init(source, target, n + m + 2)

        const nodes: number[] = io.readIntegersOfLine()
        for (let i = 0; i < n; ++i) {
          const weight: number = nodes[i]
          dinic.addEdge(i + 1, target, weight)
        }

        let answer = 0
        for (let i = 1; i <= m; ++i) {
          const [u, v, weight] = io.readIntegersOfLine()
          const x = n + i
          answer += weight
          dinic.addEdge(source, x, weight)
          dinic.addEdge(x, u, Number.MAX_SAFE_INTEGER)
          dinic.addEdge(x, v, Number.MAX_SAFE_INTEGER)
        }
        answer -= dinic.maxFlow()
        return answer
      }

      const caseDir: string = locateFixtures('codeforces/1082/g')
      const filenames: string[] = fs.readdirSync(caseDir).sort()
      for (const filename of filenames) {
        if (filename.endsWith('.in')) {
          const inputFilepath = path.join(caseDir, filename)
          const content = loadFixtures(inputFilepath)
          const output: number = solve(content)

          const answerFilepath = inputFilepath.replace(/\.in$/, '.out')
          const answer = Number(loadFixtures(answerFilepath))
          // eslint-disable-next-line jest/no-conditional-expect
          expect(output).toEqual(answer)
        }
      }

      dinic.solve(context => expect(Object.keys(context)).toMatchSnapshot())
    })
  })
})
