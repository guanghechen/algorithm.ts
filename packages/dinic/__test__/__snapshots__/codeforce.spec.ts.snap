// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeforces contest 1082 g 1`] = `
Array [
  "edgeTot",
  "dist",
  "edges",
  "G",
]
`;
